ports/unix: Add full uos.dupterm support.

andrewleech opened the pull request:
The unix port currently supports a very limited uos.dupterm implementation. 
It can only replace stdin and/or duplicate stdout. 
When using uos.dupterm to replace slot 0 it does not return a handle to the existing (posix stdio) stream.

This PR resolves these issues, preferring to use dupterm for all stdio on unix when `MICROPY_PY_OS_DUPTERM=1` is configured in the build.

With this change simple modification of the output stream is possible, for example to format logging messaging with a basic timestamp
``` python
stdio = None

class logger:
    def write(self, buf):
        global stdio        
        stdio.write(str(utime.time()))
        stdio.write(" | ")
        stdio.write(buf)

stdio = uos.dupterm(logger(), 0)
```

dlech commented on ports/unix/unix_mphal.c:
```diff
@@ -114,57 +119,85 @@ void mp_hal_stdio_mode_orig(void) {
+
+STATIC mp_uint_t unix_stdio_write(mp_obj_t self_in, const void *buf, mp_uint_t size, int *errcode) {
+    int ret;
+    MP_HAL_RETRY_SYSCALL(ret, write(STDOUT_FILENO, (const byte *)buf, size), {});
```
same here with not handling the error

dpgeorge commented:
On CPython I guess you can hook into stdout by redefining `sys.stdout`... and maybe that's how MicroPython should have done it from the start, instead of adding `uos.dupterm()` (although it'd still need an efficient multiplexer, eg `sys.stdout = micropython.stream_multiplex(sys.stdout, machine.UART(...))`, not to mention handling `sys.stdin`).

Anyway, the way it is now `sys.stdout` is *before* the dupterm multiplexer, so writing to `sys.stdout` goes to all dupterm streams (similar for reading).  And what this PR does is adds raw-stdout/stdin objects, ie they do not go via dupterm.  On bare-metal the equivalent would be a UART or USB_VCP object.

It just seems like there's a lot of complexity here with two different kinds of stdout/stdin objects (a dupterm one and a raw one).  But I don't really see a way around that.

Would it be possible to reuse the `extmod/vfs_posix_file.c` code to implement the raw stdin/stdout?  That code has compile-time dupterm support but maybe that dupterm support can be made optional at runtime (via a flag in the stdio object) so that the same code can be used for both sets of stdio objects?

andrewleech commented:
Thanks for the pointer towards `extmod/vfs_posix_file.c` - it already has the stdio object instantiated
``` c
const mp_obj_vfs_posix_file_t mp_sys_stdin_obj = {{&mp_type_textio}, STDIN_FILENO};
const mp_obj_vfs_posix_file_t mp_sys_stdout_obj = {{&mp_type_textio}, STDOUT_FILENO};
const mp_obj_vfs_posix_file_t mp_sys_stderr_obj = {{&mp_type_textio}, STDERR_FILENO};
```
I should be able to clean up this patch reusing them.

andrewleech commented on ports/unix/unix_mphal.c:
```diff
@@ -64,8 +66,11 @@ STATIC void sighandler(int signum) {
 }
 #endif
 
+int mp_interrupt_char = -1;
```
Good point. Yes I had to add this just to allow compilation with `MICROPY_PY_OS_DUPTERM` enabled.

andrewleech commented on ports/unix/mphalport.h:
```diff
@@ -103,3 +103,7 @@ enum {
 #endif
+
+#if MICROPY_PY_OS_DUPTERM
+void init_dupterm_stdio();
```
I was loath to change the name of an existing file, but I too thought it was oddly inconsistent in name.

andrewleech commented:
@dpgeorge there's a catch with using the `mp_sys_stdin_obj` etc objects above as i/o for dupterm, other things read/write to these objects - and writes to stdout/stderr are redirected to dupterm (hello recursion).

I've resolved this by adding a relatively small amount of extra logic to make a new `const mp_obj_vfs_posix_file_t mp_sys_raw_stdio_obj` which will automatically read/write to the correct fd and not go via dupterm - this new `raw_stdio` can be used for `unix_mphal.c` both with and without dupterm.

It feels like a bit of a tacked-on-the-side addition, but it does make for less duplication than I had previously.

pi-anl committed:
extmod/os_dupterm: Fix unsigned comparison warning in builtin stream path.

The MAX(0, written) comparison where written is mp_uint_t (unsigned)
triggers a -Wtype-limits warning because unsigned >= 0 is always true.
Replace with an explicit MP_STREAM_ERROR check which is the actual
intended guard.

Signed-off-by: Andrew Leech <andrew.leech@planetinnovation.com.au>

pi-anl committed:
ports/unix: Add full uos.dupterm support.

Implements proper dupterm support for the unix port using the existing
MICROPY_PY_OS_DUPTERM_BUILTIN_STREAM infrastructure.

A lightweight mp_unix_stdio_obj stream type wraps STDIN/STDOUT and is
registered as dupterm slot 0 at init. This allows os.dupterm() to manage
the default console alongside user-provided streams (e.g. for remote
REPL over sockets).

The mp_hal_stdin_rx_chr and mp_hal_stdout_tx_strn functions are
simplified to delegate to mp_os_dupterm_rx_chr/mp_os_dupterm_tx_strn.

Also provides mp_interrupt_char for dupterm keyboard interrupt detection
on non-builtin streams, and removes the stderr->dupterm echo to avoid
double output on the unix port where stderr and stdout are separate.

Signed-off-by: Andrew Leech <andrew.leech@planetinnovation.com.au>
